# RP1B_Task2

This project includes a Conda environment file to recreate the conda environment used. This can be found in the repository file **task2_env.ymal**  

## Part 1
The pipeline can be run using **compare_vcf.py** although both **mutate_genome.py** and **simulate_reads.py** have been written so that they can both be used independently.  

This is an example command used to generate any results discussed here:  
**python compare_vcf.py -i ref.fasta -c NC_004321.1 -j sp_name**  

Specific commands are listed below:  
python compare_vcf.py -i NC_037282.1.fasta -c NC_037282.1 -j plasmodium  
python compare_vcf.py -i EcoliK12-MG1655.fasta -c NC_000913.3 -j Ecoli  

All results are output into a directory:  
**results_[job_name]**  

Corresponding results can be found at the following location:  
https://jhub.climb.ac.uk/hub/user-redirect/lab/tree/shared-team/people/charlotte/task_2  
Path: shared-team/people/charlotte/task_2

All files required to run this pipeline are inside the Part_1 folder in this repository.  Some example results files are located inside the results folders within the Part_1 folder in this repository.

### mutate_genome.py 
Takes the following CLI arguments:  
**'-i'** reference genome fasta file (required, this is the genome you want to apply mutations to)  
**'-s'** number of SNPs to generate (not required, default is 300)  
**'-id'** number of indels to generate (not required, default is 20)  
**'-c'** contig ID (required, i.e. NC_004321.1)  
**'-j'** the name of the job (not required but useful to have, dictates folder naming etc. to keep results tidier when running multiple times)  

This script will take a fasta file and will randomly generate the sepcified number of SNPs and indels, returning a mutated version of the original genome.  
A fasta file titled **mutated_[contig_ID].fasta** will be generated as well as a mock vcf file (**mock_mutated_[contig_ID].vcf**) containing details for all the mutations that were applied to the original genome.  
Additional notes can be found as comments within the code.  

### test_mutate_genome.py
This uses unittest to check the functionality of the mutate_genome.py script.  
I wanted to make sure that my mutations were behaving in a way that was expected before moving on to the next stages of the project.  
This can be run from the terminal using the following command when within the same directory as mutate_genome.py: **python test_mutate_genome.py -v**  
Additional information about each test case can be found as comments within the test script.

### simulate_reads.py
Takes the following arguments:  
**'-i'** genome fasta file (required, this is the genome you want to use when simulating reads i.e. the mutated genome)  
**'-d'** required depth (not required, default is 30)  
**'-rl'** read length (not required, default is 100bp)  
**'-c'** contig ID (required, i.e. NC_004321.1)  
**'-j'** the name of the job (not required but useful to have, dictates folder naming etc. to keep results tidier when running multiple times)  

This script simulates paired reads from a fasta file and outputs two fastq files containing the reads.  These are called **[contig ID]_mutated_R1.fastq** and **[contig ID]_mutated_R2.fastq** respectively.  
Additional detail can be found within the code as comments.  

### compare_vcf.py
Takes the following arguments:  
**'-i'** reference genome fasta file (required, this is the genome you want to apply mutations to)  
**'-s'** number of SNPs to generate (not required, default is 300)  
**'-id'** number of indels to generate (not required, default is 20)  
**'-d'** required depth (not required, default is 30)  
**'-rl'** read length (not required, default is 100bp)  
**'-c'** contig ID (required, i.e. NC_004321.1)  
**'-j'** the name of the job (not required but useful to have, dictates folder naming etc. to keep results tidier when running multiple times)  

This script runs both mutate_genome.py and simulate_reads.py.  It then runs the bcftools pipeline from class and compares the variants detected with those in the mock vcf file generated by mutate_genome.py.  
The vcf from the bcftools pipeline will be called **[contig ID]_variants.vcf** and will be located within the **results_[job]** directory.  

The precision and recall of the bcftools variant caller is calculated and recorded in a file called **[contig ID]_prec_recall.txt** which is located within the results_[job] folder.  
Additional details can be found within the code as comments.

## Part 2
This part of the project relies on some files output during the previous part:  
+ Two simulated reads fastq files for both the plasmodium and E-coli genomes
+ Two fastq files containing the real reads for E-coli
+ The original genomes for both plasmodium and E-coli
+ The mock vcf files for both plasmodium and E-coli generated by mutate_genome.py (when run independently or using compare_vcf.py)  

The required files to run this pipeline are included in the Part_2 folder in this repository (apart from the fastq files which are located on CLIMB).  Some example results files are included but full results can be found on CLIMB.

https://jhub.climb.ac.uk/hub/user-redirect/lab/tree/shared-team/people/charlotte/task_2  
Path: shared-team/people/charlotte/task_2

### combine_vcf.py
Takes the following arguments:
**'-i'** reference genome fasta file (required, this is the genome you want to apply mutations to)  
**'-r1'** path for fastq file r1 of paired reads (required)  
**'-r2'** path for fastq file r2 of paired reads (required)    
**'-c'** contig ID (required, i.e. NC_004321.1)  
**'-j'** the name of the job (required, dictates folder naming etc. to keep results tidier when running multiple times)  

This script takes a reference genome and two fastq files (illumina paired reads) and uses both bcftools and snippy for variant calling.  
The resulting vcf files are then combined to create a combined vcf (**combined.[job].vcf**).  
The number of variants detected by each variant caller is also output to a file called **[job]_vc_compare.txt**.  The precision and recall of this combined method is also calculated in respect to the mock vcf created during part 1 which contains the true variants.  This file is called **[job]_combined_prec_recall.txt**.  
Additional details can be found within the code as comments.
